x/staking/keeper/liquid_stake.go

```diff
@@ -0,0 +1,158 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/iqlusioninc/liquidity-staking-module/x/staking/types"
+)
+
+// SetTotalLiquidStakedTokens stores the total outstanding tokens owned by a liquid staking provider
+func (k Keeper) SetTotalLiquidStakedTokens(ctx sdk.Context, tokens sdk.Int) {
+	store := ctx.KVStore(k.storeKey)
+
+	tokensBz, err := tokens.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	store.Set(types.TotalLiquidStakedTokensKey, tokensBz)
+}
+
+// GetTotalLiquidStakedTokens returns the total outstanding tokens owned by a liquid staking provider
+// Returns zero if the total liquid stake amount has not been initialized
+func (k Keeper) GetTotalLiquidStakedTokens(ctx sdk.Context) sdk.Int {
+	store := ctx.KVStore(k.storeKey)
+	tokensBz := store.Get(types.TotalLiquidStakedTokensKey)
+
+	if tokensBz == nil {
+		// QUESTION: Should we panic here instead?
+		// This is basically protecting against the case where we failed
+		// to bootstrap the total liquid staked in the upgrade handler
+		return sdk.ZeroInt()
+	}
+
+	var tokens sdk.Int
+	if err := tokens.Unmarshal(tokensBz); err != nil {
+		panic(err)
+	}
+
+	return tokens
+}
+
+// Check if an account is a owned by a liquid staking provider
+// This is determined by checking if the account is a 32-length module account
+func (k Keeper) AccountIsLiquidStakingProvider(ctx sdk.Context, address sdk.AccAddress) bool {
+	account := k.authKeeper.GetAccount(ctx, address)
+	_, isModuleAccount := account.(*authtypes.ModuleAccount)
```

Comment:
Can we make sure we have thorough testing on this for the following cases
- non-ICA module accounts (e.g. staking module accounts)
- interchain accounts
- regular private key accounts
- we can't test CW account afaiu on Gaia (note to self to test those before generalizing LSM to other zones)

Can you think of other types?

Reply:
I'm not sure how to differentiate a non-ICA module account from a module account in this scenario - there's no additional identifier besides an arbitrary address and name